core/src/main/java/org/apache/calcite/plan/hep/HepPlanner.java

```diff
@@ -275,7 +253,7 @@ void executeInstruction(
     LOGGER.trace("Applying rule class {}", instruction.ruleClass);
     if (instruction.ruleSet == null) {
       instruction.ruleSet = new LinkedHashSet<>();
-      for (RelOptRule rule : allRules) {
+      for (RelOptRule rule : mapDescToRule.values()) {
```

Comment:
If we already know the size of `mapDescToRule`, we can create a set with exact size, which will eliminate capacity expansion overhead and space waste when creating. Even though this is trivial update, I think it is always a good manner to create collection in such way if possible.
```
instruction.ruleSet = Sets.newHashSetWithExpectedSize(mapDescToRule.size());
```

Reply:
Yes, will do.

Reply:
It turns out the size instruction.ruleSet is undetermined, we don't know the exact size. So I will leave as it is.

Reply:
I agree, there is "if" check to determine what to be added. I didn't find that out clearly, sorry about the false alert.